File: src/octopus_scraper/metrics.py
"""Prometheus metrics for OctopusScraper."""

import time
from importlib.metadata import PackageNotFoundError, version
from typing import Any, Optional

from prometheus_client import (
    CollectorRegistry,
    Counter,
    Gauge,
    Histogram,
    Info,
    generate_latest,
)


class OctopusMetrics:
    """Own and update the service Prometheus metric registry."""

    TASK_DURATION_BUCKETS = (1, 5, 15, 30, 60, 120, 300, 600, 1200, 2400)
    EXTERNAL_DURATION_BUCKETS = (0.1, 0.5, 1, 2.5, 5, 10, 30, 60, 120, 300, 1200)
    ITEM_BUCKETS = (0, 1, 5, 10, 25, 50, 100, 250, 500, 1000)
    DEPENDENCIES = frozenset({"rss", "notion", "llm"})

    def __init__(self, registry: Optional[CollectorRegistry] = None) -> None:
        self.registry = registry or CollectorRegistry()
        self._start_time = time.time()

        self.build_info = Info(
            "octopus_build",
            "OctopusScraper build information.",
            registry=self.registry,
        )
        self.service_start_time = Gauge(
            "octopus_service_start_time_seconds",
            "Unix timestamp when the service process started.",
            registry=self.registry,
        )
        self.service_uptime = Gauge(
            "octopus_service_uptime_seconds",
            "Service process uptime in seconds.",
            registry=self.registry,
        )

        self.tasks_submitted = Counter(
            "octopus_tasks_submitted_total",
            "Task attempts accepted by the task manager.",
            registry=self.registry,
        )
        self.tasks_completed = Counter(
            "octopus_tasks_completed_total",
            "Task attempts completed successfully.",
            registry=self.registry,
        )
        self.tasks_failed = Counter(
            "octopus_tasks_failed_total",
            "Task attempts that failed.",
            registry=self.registry,
        )
        self.task_retries = Counter(
            "octopus_task_retries_total",
            "Retry task attempts successfully re-enqueued.",
            registry=self.registry,
        )
        self.tasks_cancelled = Counter(
            "octopus_tasks_cancelled_total",
            "Task attempts cancelled before completion.",
            registry=self.registry,
        )
        self.tasks_running = Gauge(
            "octopus_tasks_running",
            "Task attempts currently running.",
            registry=self.registry,
        )
        self.tasks_queued = Gauge(
            "octopus_tasks_queued",
            "Task attempts currently waiting in the queue.",
            registry=self.registry,
        )
        self.queue_capacity = Gauge(
            "octopus_task_queue_capacity",
            "Maximum number of queued task attempts.",
            registry=self.registry,
        )
        self.worker_capacity = Gauge(
            "octopus_task_worker_capacity",
            "Maximum number of concurrent task workers.",
            registry=self.registry,
        )
        self.task_duration = Histogram(
            "octopus_task_duration_seconds",
            "Task attempt execution duration.",
            buckets=self.TASK_DURATION_BUCKETS,
            registry=self.registry,
        )
        self.task_items_fetched = Histogram(
            "octopus_task_items_fetched",
            "Items fetched by a completed task attempt.",
            buckets=self.ITEM_BUCKETS,
            registry=self.registry,
        )

        self.upload_items = Counter(
            "octopus_upload_items_total",
            "Items processed by upload operations.",
            labelnames=("outcome",),
            registry=self.registry,
        )
        self.upload_batch_size = Histogram(
            "octopus_upload_batch_items",
            "Items included in an upload operation.",
            buckets=self.ITEM_BUCKETS,
            registry=self.registry,
        )

        self.config_healthy = Gauge(
            "octopus_config_healthy",
            "Whether the current scraper configuration is healthy.",
            registry=self.registry,
        )
        self.config_refresh_successes = Counter(
            "octopus_config_refresh_success_total",
            "Successful configuration load or refresh operations.",
            registry=self.registry,
        )
        self.config_refresh_failures = Counter(
            "octopus_config_refresh_failure_total",
            "Failed configuration load or refresh operations.",
            registry=self.registry,
        )
        self.config_last_success = Gauge(
            "octopus_config_last_success_timestamp_seconds",
            "Unix timestamp of the last successful configuration load or refresh.",
            registry=self.registry,
        )

        self.external_requests = Counter(
            "octopus_external_requests_total",
            "External dependency operations.",
            labelnames=("dependency",),
            registry=self.registry,
        )
        self.external_failures = Counter(
            "octopus_external_request_failures_total",
            "Failed external dependency operations.",
            labelnames=("dependency",),
            registry=self.registry,
        )
        self.external_duration = Histogram(
            "octopus_external_request_duration_seconds",
            "External dependency operation duration.",
            labelnames=("dependency",),
            buckets=self.EXTERNAL_DURATION_BUCKETS,
            registry=self.registry,
        )

        try:
            package_version = version("octopus-scraper")
        except PackageNotFoundError:
            package_version = "unknown"
        self.build_info.info({"version": package_version})
        self.set_service_start_time(self._start_time)
        self.tasks_running.set(0)
        self.tasks_queued.set(0)
        self.config_healthy.set(0)

    def set_service_start_time(self, timestamp: Optional[float] = None) -> None:
        """Set the process start timestamp used by service metrics."""
        self._start_time = timestamp or time.time()
        self.service_start_time.set(self._start_time)

    def configure_task_manager(self, queue_capacity: int, workers: int) -> None:
        """Set task manager capacities."""
        self.queue_capacity.set(queue_capacity)
        self.worker_capacity.set(workers)

    def set_task_state(self, queued: int, running: int) -> None:
        """Update current task queue and worker usage."""
        self.tasks_queued.set(queued)
        self.tasks_running.set(running)

    def record_task_completed(self, duration: float, items_fetched: int) -> None:
        """Record a successful task attempt."""
        self.tasks_completed.inc()
        self.task_duration.observe(duration)
        self.task_items_fetched.observe(items_fetched)

    def record_task_failed(self, duration: float) -> None:
        """Record a failed task attempt."""
        self.tasks_failed.inc()
        self.task_duration.observe(duration)

    def record_upload(self, requested: int, processed: int, failed: int) -> None:
        """Record one upload operation using the storage result contract."""
        self.upload_batch_size.observe(requested)
        if processed:
            self.upload_items.labels(outcome="processed").inc(processed)
        if failed:
            self.upload_items.labels(outcome="failed").inc(failed)

    def record_config_refresh(self, success: bool) -> None:
        """Record a configuration load or refresh outcome."""
        self.config_healthy.set(1 if success else 0)
        if success:
            self.config_refresh_successes.inc()
            self.config_last_success.set_to_current_time()
        else:
            self.config_refresh_failures.inc()

    def record_external_request(
        self, dependency: str, duration: float, success: bool
    ) -> None:
        """Record one high-level external dependency operation."""
        if dependency not in self.DEPENDENCIES:
            raise ValueError(f"Unsupported metrics dependency: {dependency}")
        self.external_requests.labels(dependency=dependency).inc()
        self.external_duration.labels(dependency=dependency).observe(duration)
        if not success:
            self.external_failures.labels(dependency=dependency).inc()

    def refresh_app_state(self, app: Any) -> None:
        """Refresh scrape-time gauges from stable service state."""
        self.service_uptime.set(max(0, time.time() - self._start_time))
        if hasattr(app.ctx, "config_manager"):
            status = app.ctx.config_manager.get_status()
            self.config_healthy.set(1 if status.is_healthy else 0)

    def render(self) -> bytes:
        """Render the registry in Prometheus exposition format."""
        return generate_latest(self.registry)


metrics = OctopusMetrics()

File: internal/observability/metrics.go
package observability

import (
	"fmt"
	"net/http"
	"time"

	"github.com/Oligarchy-with-DeamoV/OctopusScraper/internal/task"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

var _ task.Observer = (*Metrics)(nil)

type Metrics struct {
	registry          *prometheus.Registry
	startedAt         time.Time
	buildInfo         *prometheus.GaugeVec
	serviceStart      prometheus.Gauge
	serviceUptime     prometheus.Gauge
	tasksSubmitted    prometheus.Counter
	tasksCompleted    prometheus.Counter
	tasksFailed       prometheus.Counter
	taskRetries       prometheus.Counter
	tasksCancelled    prometheus.Counter
	tasksRunning      prometheus.Gauge
	tasksQueued       prometheus.Gauge
	queueCapacity     prometheus.Gauge
	workerCapacity    prometheus.Gauge
	taskDuration      prometheus.Histogram
	taskItemsFetched  prometheus.Histogram
	uploadItems       *prometheus.CounterVec
	uploadBatch       prometheus.Histogram
	configHealthy     prometheus.Gauge
	configSuccesses   prometheus.Counter
	configFailures    prometheus.Counter
	configLastSuccess prometheus.Gauge
	externalRequests  *prometheus.CounterVec
	externalFailures  *prometheus.CounterVec
	externalDuration  *prometheus.HistogramVec
}

func NewMetrics(version string) *Metrics {
	registry := prometheus.NewRegistry()
	metrics := &Metrics{
		registry:  registry,
		startedAt: time.Now(),
		buildInfo: prometheus.NewGaugeVec(prometheus.GaugeOpts{
			Name: "octopus_build_info",
			Help: "OctopusScraper build information.",
		}, []string{"version"}),
		serviceStart: prometheus.NewGauge(prometheus.GaugeOpts{
			Name: "octopus_service_start_time_seconds",
			Help: "Unix timestamp when the service process started.",
		}),
		serviceUptime: prometheus.NewGauge(prometheus.GaugeOpts{
			Name: "octopus_service_uptime_seconds",
			Help: "Service process uptime in seconds.",
		}),
		tasksSubmitted: prometheus.NewCounter(prometheus.CounterOpts{
			Name: "octopus_tasks_submitted_total",
			Help: "Task attempts accepted by the task manager.",
		}),
		tasksCompleted: prometheus.NewCounter(prometheus.CounterOpts{
			Name: "octopus_tasks_completed_total",
			Help: "Task attempts completed successfully.",
		}),
		tasksFailed: prometheus.NewCounter(prometheus.CounterOpts{
			Name: "octopus_tasks_failed_total",
			Help: "Task attempts that failed.",
		}),
		taskRetries: prometheus.NewCounter(prometheus.CounterOpts{
			Name: "octopus_task_retries_total",
			Help: "Retry task attempts successfully re-enqueued.",
		}),
		tasksCancelled: prometheus.NewCounter(prometheus.CounterOpts{
			Name: "octopus_tasks_cancelled_total",
			Help: "Task attempts cancelled before completion.",
		}),
		tasksRunning: prometheus.NewGauge(prometheus.GaugeOpts{
			Name: "octopus_tasks_running",
			Help: "Task attempts currently running.",
		}),
		tasksQueued: prometheus.NewGauge(prometheus.GaugeOpts{
			Name: "octopus_tasks_queued",
			Help: "Task attempts currently waiting in the queue.",
		}),
		queueCapacity: prometheus.NewGauge(prometheus.GaugeOpts{
			Name: "octopus_task_queue_capacity",
			Help: "Maximum number of queued task attempts.",
		}),
		workerCapacity: prometheus.NewGauge(prometheus.GaugeOpts{
			Name: "octopus_task_worker_capacity",
			Help: "Maximum number of concurrent task workers.",
		}),
		taskDuration: prometheus.NewHistogram(prometheus.HistogramOpts{
			Name:    "octopus_task_duration_seconds",
			Help:    "Task attempt execution duration.",
			Buckets: []float64{1, 5, 15, 30, 60, 120, 300, 600, 1200, 2400},
		}),
		taskItemsFetched: prometheus.NewHistogram(prometheus.HistogramOpts{
			Name:    "octopus_task_items_fetched",
			Help:    "Items fetched by a completed task attempt.",
			Buckets: []float64{0, 1, 5, 10, 25, 50, 100, 250, 500, 1000},
		}),
		uploadItems: prometheus.NewCounterVec(prometheus.CounterOpts{
			Name: "octopus_upload_items_total",
			Help: "Items processed by upload operations.",
		}, []string{"outcome"}),
		uploadBatch: prometheus.NewHistogram(prometheus.HistogramOpts{
			Name:    "octopus_upload_batch_items",
			Help:    "Items included in an upload operation.",
			Buckets: []float64{0, 1, 5, 10, 25, 50, 100, 250, 500, 1000},
		}),
		configHealthy: prometheus.NewGauge(prometheus.GaugeOpts{
			Name: "octopus_config_healthy",
			Help: "Whether the current scraper configuration is healthy.",
		}),
		configSuccesses: prometheus.NewCounter(prometheus.CounterOpts{
			Name: "octopus_config_refresh_success_total",
			Help: "Successful configuration load or refresh operations.",
		}),
		configFailures: prometheus.NewCounter(prometheus.CounterOpts{
			Name: "octopus_config_refresh_failure_total",
			Help: "Failed configuration load or refresh operations.",
		}),
		configLastSuccess: prometheus.NewGauge(prometheus.GaugeOpts{
			Name: "octopus_config_last_success_timestamp_seconds",
			Help: "Unix timestamp of the last successful configuration load or refresh.",
		}),
		externalRequests: prometheus.NewCounterVec(prometheus.CounterOpts{
			Name: "octopus_external_requests_total",
			Help: "External dependency operations.",
		}, []string{"dependency"}),
		externalFailures: prometheus.NewCounterVec(prometheus.CounterOpts{
			Name: "octopus_external_request_failures_total",
			Help: "Failed external dependency operations.",
		}, []string{"dependency"}),
		externalDuration: prometheus.NewHistogramVec(prometheus.HistogramOpts{
			Name:    "octopus_external_request_duration_seconds",
			Help:    "External dependency operation duration.",
			Buckets: []float64{0.1, 0.5, 1, 2.5, 5, 10, 30, 60, 120, 300, 1200},
		}, []string{"dependency"}),
	}
	registry.MustRegister(
		metrics.buildInfo,
		metrics.serviceStart,
		metrics.serviceUptime,
		metrics.tasksSubmitted,
		metrics.tasksCompleted,
		metrics.tasksFailed,
		metrics.taskRetries,
		metrics.tasksCancelled,
		metrics.tasksRunning,
		metrics.tasksQueued,
		metrics.queueCapacity,
		metrics.workerCapacity,
		metrics.taskDuration,
		metrics.taskItemsFetched,
		metrics.uploadItems,
		metrics.uploadBatch,
		metrics.configHealthy,
		metrics.configSuccesses,
		metrics.configFailures,
		metrics.configLastSuccess,
		metrics.externalRequests,
		metrics.externalFailures,
		metrics.externalDuration,
	)
	metrics.buildInfo.WithLabelValues(version).Set(1)
	metrics.serviceStart.Set(float64(metrics.startedAt.Unix()))
	return metrics
}

func (m *Metrics) Handler() http.Handler {
	handler := promhttp.HandlerFor(m.registry, promhttp.HandlerOpts{})
	return http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
		m.RefreshUptime()
		handler.ServeHTTP(writer, request)
	})
}

func (m *Metrics) RefreshUptime() {
	m.serviceUptime.Set(time.Since(m.startedAt).Seconds())
}

func (m *Metrics) Configure(queueCapacity, workerCapacity int) {
	m.queueCapacity.Set(float64(queueCapacity))
	m.workerCapacity.Set(float64(workerCapacity))
}

func (m *Metrics) Submitted() { m.tasksSubmitted.Inc() }
func (m *Metrics) Retried()   { m.taskRetries.Inc() }
func (m *Metrics) Cancelled() { m.tasksCancelled.Inc() }
func (m *Metrics) Failed(duration time.Duration) {
	m.tasksFailed.Inc()
	m.taskDuration.Observe(duration.Seconds())
}
func (m *Metrics) Completed(duration time.Duration, itemsFetched int) {
	m.tasksCompleted.Inc()
	m.taskDuration.Observe(duration.Seconds())
	m.taskItemsFetched.Observe(float64(itemsFetched))
}
func (m *Metrics) State(queued, running int) {
	m.tasksQueued.Set(float64(queued))
	m.tasksRunning.Set(float64(running))
}

func (m *Metrics) RecordConfig(success bool) {
	m.SetConfigHealth(success)
	if success {
		m.configSuccesses.Inc()
		m.configLastSuccess.SetToCurrentTime()
		return
	}
	m.configFailures.Inc()
}

func (m *Metrics) SetConfigHealth(healthy bool) {
	if healthy {
		m.configHealthy.Set(1)
		return
	}
	m.configHealthy.Set(0)
}

func (m *Metrics) RecordExternal(dependency string, duration time.Duration, success bool) error {
	switch dependency {
	case "rss", "notion", "llm":
	default:
		return fmt.Errorf("unsupported metrics dependency: %s", dependency)
	}
	m.externalRequests.WithLabelValues(dependency).Inc()
	m.externalDuration.WithLabelValues(dependency).Observe(duration.Seconds())
	if !success {
		m.externalFailures.WithLabelValues(dependency).Inc()
	}
	return nil
}

func (m *Metrics) RecordUpload(requested, processed, failed int) {
	m.uploadBatch.Observe(float64(requested))
	if processed > 0 {
		m.uploadItems.WithLabelValues("processed").Add(float64(processed))
	}
	if failed > 0 {
		m.uploadItems.WithLabelValues("failed").Add(float64(failed))
	}
}
